app/server/api/search_api.py

```diff
@@ -0,0 +1,124 @@
+        Return Value:
+            Results object, similar to the existing transfer_accounts and credit_transfers API return values
+        """
+        search_string = request.args.get('search_string') or ''
+
+        # Valid search types are: `transfer_accounts` and `credit_transfers`
+        search_type = request.args.get('search_type') or 'transfer_accounts'
+        if search_type not in ['transfer_accounts', 'credit_transfers']:
+            response_object = {
+                'message': 'Invalid search_type \'{}\'. Please use type \'transfer_accounts\' or \'credit_transfers\''.format(search_type),
+            }
+            return make_response(jsonify(response_object)), 400
+
+        # Note: Using tsquery wildcards here. Good docs of them here:
+        # https://www.postgresql.org/docs/current/datatype-textsearch.html#DATATYPE-TSQUERY
+        # 'Fran deRoo' -> 'Fran:* | deRoo:*'
+        # Matches strings like "Francine deRoos"
+        # Will also match "Michiel deRoos" because of the or clause, but this will be ranked lower
+        search_terms = search_string.strip().split(' ')
+        tsquery = ':* | '.join(search_terms)+':*'
+
+        # Return everything if the search string is empty
+        if search_string == '':
+            if search_type == 'transfer_accounts':
+                final_query = TransferAccount.query.filter(TransferAccount.is_ghost != True)\
+                    .execution_options(show_all=True)
+                transfer_accounts, total_items, total_pages = paginate_query(final_query, TransferAccount)
+                result = transfer_accounts_schema.dump(transfer_accounts)
+                data = { 'transfer_accounts': result.data }
+            else:
+                final_query = CreditTransfer.query.filter()\
+                    .execution_options(show_all=True)
+                credit_transfers, total_items, total_pages = paginate_query(final_query, CreditTransfer)
+                result = credit_transfers_schema.dump(credit_transfers)
+                data = { 'credit_transfers': result.data }
+
+        else:
+            # First get users who match search string
+            user_search_result = db.session.query(
+                db.distinct(SearchView.id),
+                SearchView,
+                # This ugly (but functional) multi-tscolumn ranking is a modified from Ben Smithgall's blog post
+                # https://www.codeforamerica.org/blog/2015/07/02/multi-table-full-text-search-with-postgres-flask-and-sqlalchemy/
+                db.func.max(db.func.full_text.ts_rank(
+                    db.func.setweight(db.func.coalesce(SearchView.tsv_email, ''), 'D')\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_phone, ''), 'A'))\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_first_name, ''), 'B'))\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_last_name, ''), 'B')),
```

Comment:
I feel like the ability to also search for the following items would be very useful...

- Public Serial Number
- Location
- Blockchain Address
- And if possible, _any other custom attribute_ (i.e. Bio)

Not sure how hard that would be to add though!

Reply:
Added serial/address/location. I think indexing custom attributes would be a bit trickier. Should we spin that off into another ticket?

Reply:
Yeah, I think keep custom attribute separate if it's to hard